src/gpl/src/mbff.cpp

```diff
@@ -0,0 +1,1058 @@
+            cp_model.AddLessOrEqual(b_slot, 1);
+        }
+    }
+
+    for (int i = 0; i < num_flops; i++) {
+        operations_research::sat::LinearExpr b_flop;
+        for (int j = 0; j < num_trays; j++) {
+            for (int k = 0; k < static_cast<int>(trays[j].slots.size()); k++)
+                if (trays[j].cand[k] == i) {
+                    b_flop += B[i][j][k];
+                }
+        }
+        cp_model.AddLessOrEqual(1, b_flop);
+        cp_model.AddLessOrEqual(b_flop, 1);
+    }
+    // END CONSTRAINTS FOR B
+
+    // CONSTRAINTS FOR E
+    std::vector<operations_research::sat::BoolVar> e(num_trays);
+    for (int i = 0; i < num_trays; i++) {
+        e[i] = cp_model.NewBoolVar();
+    }
+
+    for (int i = 0; i < num_trays; i++) {
+        operations_research::sat::LinearExpr slots_used;
+        for (int j = 0; j < static_cast<int>(trays[i].slots.size()); j++) {
+            for (int k = 0; k < num_flops; k++)
+                if (trays[i].cand[j] == k) {
+                    cp_model.AddLessOrEqual(0, e[i] - B[k][i][j]);
+                    slots_used += B[k][i][j];
+                }
+        }
+        cp_model.AddLessOrEqual(0, slots_used - e[i]);
+    }
+    // END CONSTRAINTS FOR E
+
+    // calculate the cost of each tray
+    std::vector<float> w(num_trays);
+    for (int i = 0; i < num_trays; i++) {
+        int bit_idx = 0;
+        for (int j = 0; j < 6; j++) {
+            if (GetBitCnt(j) == static_cast<int>(trays[i].slots.size()))
+                bit_idx = j;
+        }
+        if (GetBitCnt(bit_idx) == 1)
+            w[i] = 1.00;
+        else
+            w[i] = ((float)GetBitCnt(bit_idx)) *
+                   (RATIOS[bit_idx] - GetBitCnt(bit_idx) * 0.0015);
+    }
+
+    /*
+            - DoubleLinearExpr support double coefficients
+            - can only be used for the objective function
+    */
+
+    operations_research::sat::DoubleLinearExpr obj;
+
+    // add B
+    for (int i = 0; i < num_flops; i++) {
+        for (int j = 0; j < num_trays; j++) {
+            for (int k = 0; k < static_cast<int>(trays[j].slots.size()); k++)
+                if (trays[j].cand[k] == i) {
+                    obj.AddTerm(d_x[i][j][k], 0.01);
+                    obj.AddTerm(d_y[i][j][k], 0.01);
+                }
+        }
+    }
+
+    // add Z
+    for (int i = 0; i < num_paths; i++) {
+        obj.AddTerm(z_x[i], BETA * 0.01);
+        obj.AddTerm(z_y[i], BETA * 0.01);
+    }
+
+    // add E
+    for (int i = 0; i < num_trays; i++) {
+        obj.AddTerm(e[i], ALPHA * w[i]);
+    }
+
+    cp_model.Minimize(obj);
+
+    const operations_research::sat::CpSolverResponse response =
+        operations_research::sat::Solve(cp_model.Build());
+    if (response.status() ==
+        operations_research::sat::CpSolverStatus::INFEASIBLE) {
+        std::cout << "No solution found.\n";
+        return;
+    }
+
+    std::cout << "Total = " << (response.objective_value()) << "\n";
```

Comment:
Use the logger rather than cout (throughout).  Consider whether this is for developer use (debugPrint) or the user (info).  See https://github.com/The-OpenROAD-Project/OpenROAD/blob/master/docs/contrib/Logger.md